### PR TITLE
Use sha256 to hash StepFunctions trace id and manually set `_dd.p.tid`

### PR DESCRIPTION
**Pull request by kimi-p**
### What does this PR do?
- Use sha256 to hash StepFunctions trace context. 
- For 128 bits `trace_id`:
  - The 1st and 65th bits should be `0`.
  - The 1st 64 bits will be converted to base-16 to `_dd.p.tid` tag.
  - The 2nd 64 bits will be put into trace_id in base-10
- For `parent_id` (which is `aws.stepfunctions.lambda` span's span_id in this case):
  - The highest 64 bits will be converted to base-10.
  - The 1st bit should be `0`.
- The trace_id on the APM UI will now be in base-16 instead of base-10.
- [Related PR](https://github.com/DataDog/datadog-lambda-python/pull/320) from the past that uses md5 to hash.

### Motivation
To support 128 bits trace IDs and to avoid showing up in security vulnerability scans, we are upgrading the hashing method (md5) to sha256.

### Testing Guidelines
Tested in staging. The example trace is from a Step Function that triggers two Lambdas. One using NodeJS; another using Python. 
- [example trace](https://dd.datad0g.com/functions?cloud=aws&entity_view=lambda_functions&fromUser=false&graphType=flamegraph&panel_end=1716483562551&panel_paused=false&panel_start=1716479962551&shouldShowLegend=true&sort=time&sp=%5B%7B%22p%22%3A%7B%22entityId%22%3A%22aws-lambda-functions%2Bkimi-test-step-functions-linking%2Bsa-east-1%2B425362996713%22%7D%2C%22i%22%3A%22lambda-panel%22%7D%2C%7B%22p%22%3A%7B%22traceID%22%3A%2265ed9415f24950ca6af10dc3ae25a85a%22%2C%22selectedSpanID%22%3A%222432504545734628660%22%7D%2C%22i%22%3A%22trace-panel%22%7D%5D&spanID=2432504545734628660&spanViewType=metadata&text_search=kimi&traceID=65ed9415f24950ca6af10dc3ae25a85a&view=spans&start=1716479482566&end=1716483082566&paused=false)
image

### Types of Changes

- [ ] Bug fix
- [x] New feature
- [x] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: tests/test_tracing.py [hunk 1]
```diff
@@ -1992,19 +1992,26 @@ def test_mark_trace_as_error_for_5xx_responses_sends_error_metric_and_set_error_
...
-    def test_deterministic_m5_hash__result_the_same_as_backend(self):
+    def test_deterministic_m5_hash__result_the_same_as_backend_1(self):
         result = _deterministic_md5_hash(
-            "arn:aws:states:sa-east-1:601427271234:express:DatadogStateMachine:acaf1a67-336a-e854-1599-2a627eb2dd8a"
-            ":c8baf081-31f1-464d-971f-70cb17d01111#step-one#2022-12-08T21:08:19.224Z"
+            "arn:aws:states:sa-east-1:425362996713:stateMachine:MyStateMachine-b276uka1j"
+            "#lambda#1", 64
         )
-        self.assertEqual(8034507082463708833, result)
+        self.assertEqual(3711631873188331089, result)
```
Comment (kimi-p) [inline tests/test_tracing.py:2003]: Same hash as logs-backend
https://github.com/DataDog/logs-backend/pull/71616/files#diff-820e0a1dd4f0f97815ac54f993c08b3c44c67dd26a312968c214f16ab73494c0R12-R15

File: tests/test_tracing.py [hunk 2]
```diff
@@ -1992,19 +1992,26 @@ def test_mark_trace_as_error_for_5xx_responses_sends_error_metric_and_set_error_
...
+
+    def test_deterministic_m5_hash__result_the_same_as_backend_2(self):
+        result = _deterministic_md5_hash(
+            "arn:aws:states:sa-east-1:425362996713:stateMachine:MyStateMachine-b276uka1j"
+            "#lambda#2", 64
+        )
+        self.assertEqual(5759173372325510050, result)
```
Comment (kimi-p) [inline tests/test_tracing.py:2010]: Same hash as logs-backend.
https://github.com/DataDog/logs-backend/pull/71616/files#diff-820e0a1dd4f0f97815ac54f993c08b3c44c67dd26a312968c214f16ab73494c0R17-R21

**Review by agocs — APPROVED** [review]

File: datadog_lambda/tracing.py [hunk 3]
```diff
@@ -354,14 +356,16 @@ def extract_context_from_kinesis_event(event, lambda_context):
...
+def _deterministic_sha256_hash(s: str, part: str) -> (int, int):
+    sha256_hash = hashlib.sha256(s.encode()).hexdigest()
+
+    # First two chars is '0b'. zfill to ensure 256 bits, but we only care about the first 128 bits
+    binary_hash = bin(int(sha256_hash, 16))[2:].zfill(256)
+    if part == HIGHER_64_BITS:
+        updated_binary_hash = "0" + binary_hash[1:64]
+    else:
+        updated_binary_hash = "0" + binary_hash[65:128]
+    result = int(updated_binary_hash, 2)
```
Comment (nine5two7) [inline datadog_lambda/tracing.py:368]: Here is an base 2 integer. But later, it is converted to a base 10 integer somewhere?

Comment (kimi-p) [reply datadog_lambda/tracing.py:368]: I think we are returning in both binary and format.
- [returning binary](https://github.com/DataDog/datadog-lambda-python/blob/51b272dd04d250aa4aa1ede77b7ebf00365dab48/datadog_lambda/tracing.py#L386-L390)
- [convert into hex](https://github.com/DataDog/datadog-lambda-python/blob/51b272dd04d250aa4aa1ede77b7ebf00365dab48/datadog_lambda/tracing.py#L400-L402)

**Review by nine5two7 — APPROVED** [review]
LGTM

**Review by duncanista — APPROVED** [review]

**Pull request merged**